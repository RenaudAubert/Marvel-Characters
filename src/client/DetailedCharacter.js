import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { FavButton, LinkType } from './layout';

export default class DetailedCharacter extends Component {
  constructor(props) {
    super(props);

    this.state = {character: {}};
  }

  // Get data from Marvel API if component is mounted
  componentDidMount() {
    const { match, shouldFetch } = this.props;

    // Prevents fetching if character was found in characters array in App
    if (shouldFetch) {
      // Fetch a single character using id
      fetch(`/api/characters/${match.params.id}`)
      .then(res => res.json())
      .then(apiResult => {
        if (apiResult.data && apiResult.data.results.length) {
          apiResult.data.results[0].favorite = false;
          this.setState({ character: apiResult.data.results[0] });
        }
      });
    }
  }

  // Render each character card with image, name and description
  render() {
    let { character, shouldFetch, onFavClicked } = this.props;
    // Assign value returned by Marvel API if character was not found in App
    if (shouldFetch) {
      character = this.state.character;
    }
    let returnedDOM = (
      <p className="text-center text-danger">
        <strong>No character found</strong>
      </p>
    );

    // If object is empty
    if (Object.keys(character).length !== 0) {
      const imgURL = character.thumbnail ? `${character.thumbnail.path}/landscape_large.${character.thumbnail.extension}` : '';
      // Only list the 4 first comics returned
      const comicsList = (type, length) => {
        const list = character[type].items.slice(0, length).map(comic => (
          <li key={comic.name}>{comic.name}</li>
        ));

        return <ul>{list}</ul>;
      };

      returnedDOM = (
        <div>
          <div className="row">
            <div className="col-lg-5">
              <img className="card-img-left flex-auto d-none d-md-block" src={imgURL} alt="" />
            </div>
            <div className="col-lg-7">
              <h5>
                <LinkType character={character} type="detail" text={character.name} />
                <FavButton characterId={character.id} isFav={character.favorite} onFavClicked={onFavClicked} />
              </h5>
              <h6>Description</h6>
              {character.description ?
                <p className="card-text">{character.description}</p> :
                <p className="card-text text-danger">{character.description}</p>
              }
              <div className="row">
                <div className="col-sm-6">
                  <h6>
                    Comics
                    <LinkType character={character} type="comiclink" text={` (${character.comics.available})`} />
                  </h6>
                  {
                    character.comics.items.slice(0, 4).map(comic => (
                      <li key={comic.name}>{comic.name}</li>
                    ))
                  }
                  <LinkType character={character} type="comiclink" text="Details" />
                </div>
                <div className="col-sm-6">
                  <h6>Series</h6>
                  {
                    character.series.items.slice(0, 4).map(serie => (
                      <li key={serie.name}>{serie.name}</li>
                    ))
                  }
                </div>
              </div>
            </div>
          </div>
        </div>
      );
    }

    return (
      <div>
        <Link to="/" className="pl-2">
          <i className="far fa-arrow-alt-circle-left fa-3x text-secondary" />
        </Link>
        <div className="container pt-2">
          {returnedDOM}
        </div>
      </div>
  );
}
}

DetailedCharacter.propTypes = {
  character: PropTypes.shape({
    id: PropTypes.number,
    description: PropTypes.string,
    name: PropTypes.string,
    thumbnail: PropTypes.object,
    favorite: PropTypes.bool
  }),
  onFavClicked: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.node,
    }).isRequired,
  }).isRequired,
  shouldFetch: PropTypes.bool
};

DetailedCharacter.defaultProps = {
  character: {},
  shouldFetch: false
};
