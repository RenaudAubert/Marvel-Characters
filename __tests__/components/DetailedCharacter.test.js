import React from 'react';
import { shallow, mount  } from 'enzyme';
import { Link, MemoryRouter } from "react-router-dom";
import toJson from 'enzyme-to-json';
import apiResult from '../__mocks__/characters';

// Component to be tested
import DetailedCharacter from '../../src/client/DetailedCharacter';

const character = apiResult.data.results[0];
const match = { params: { id: 0 } };

beforeEach(() => {
  global.fetch = jest.fn().mockImplementation(() => {
    const p = new Promise((resolve, reject) => {
      resolve(character);
    });
    return p;
  });
});

describe('<DetailedCharacter />', () => {
  const mockOnClick = jest.fn();

  describe('render()', () => {
    test('renders the DetailedCharacter component with character', () => {
      const wrapper = shallow(
        <DetailedCharacter
          character={character}
          onFavClicked={mockOnClick}
          match={match}
          shouldFetch={false}
        />
      );
      expect(wrapper.contains(<p className="text-center text-danger"><strong>No character found</strong></p>)).toEqual(false);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('renders the DetailedCharacter component without character', () => {
      const wrapper = shallow(
        <DetailedCharacter
          character={{}}
          onFavClicked={mockOnClick}
          match={match}
          shouldFetch={false}
        />
      );
      expect(wrapper.contains(<p className="text-center text-danger"><strong>No character found</strong></p>)).toEqual(true);
    });

    describe('fetching data after mount', () => {
      test('succesfully fetching data', () => {
        const wrapper = mount(
          <MemoryRouter>
            <DetailedCharacter
              character={{}}
              onFavClicked={mockOnClick}
              match={match}
              shouldFetch={true}
            />
          </MemoryRouter>
        );
        wrapper.update();
        expect(wrapper.find(DetailedCharacter).state().character).toEqual(character);
      });
    });
  });
});
