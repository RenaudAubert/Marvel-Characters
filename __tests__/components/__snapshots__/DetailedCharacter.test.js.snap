// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailedCharacter /> render() renders the DetailedCharacter component with character 1`] = `
<div>
  <Link
    className="pl-2"
    replace={false}
    to="/"
  >
    <i
      className="far fa-arrow-alt-circle-left fa-3x text-secondary"
    />
  </Link>
  <div
    className="container pt-2"
  >
    <div>
      <div
        className="row"
      >
        <div
          className="col-lg-5"
        >
          <img
            alt=""
            className="card-img-left flex-auto d-none d-md-block"
            src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784/landscape_large.jpg"
          />
        </div>
        <div
          className="col-lg-7"
        >
          <h5>
            <LinkType
              character={
                Object {
                  "comics": Object {
                    "available": 12,
                    "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/comics",
                    "items": Array [
                      Object {
                        "name": "Avengers: The Initiative (2007) #14",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/21366",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #14 (SPOTLIGHT VARIANT)",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/24571",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #15",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/21546",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #16",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/21741",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #17",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/21975",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #18",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/22299",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #18 (ZOMBIE VARIANT)",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/22300",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #19",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/22506",
                      },
                      Object {
                        "name": "Deadpool (1997) #44",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/8500",
                      },
                      Object {
                        "name": "Marvel Premiere (1972) #35",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/10223",
                      },
                      Object {
                        "name": "Marvel Premiere (1972) #36",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/10224",
                      },
                      Object {
                        "name": "Marvel Premiere (1972) #37",
                        "resourceURI": "http://gateway.marvel.com/v1/public/comics/10225",
                      },
                    ],
                    "returned": 12,
                  },
                  "description": "",
                  "events": Object {
                    "available": 1,
                    "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/events",
                    "items": Array [
                      Object {
                        "name": "Secret Invasion",
                        "resourceURI": "http://gateway.marvel.com/v1/public/events/269",
                      },
                    ],
                    "returned": 1,
                  },
                  "id": 1011334,
                  "modified": "2014-04-29T14:18:17-0400",
                  "name": "3-D Man",
                  "resourceURI": "http://gateway.marvel.com/v1/public/characters/1011334",
                  "series": Object {
                    "available": 3,
                    "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/series",
                    "items": Array [
                      Object {
                        "name": "Avengers: The Initiative (2007 - 2010)",
                        "resourceURI": "http://gateway.marvel.com/v1/public/series/1945",
                      },
                      Object {
                        "name": "Deadpool (1997 - 2002)",
                        "resourceURI": "http://gateway.marvel.com/v1/public/series/2005",
                      },
                      Object {
                        "name": "Marvel Premiere (1972 - 1981)",
                        "resourceURI": "http://gateway.marvel.com/v1/public/series/2045",
                      },
                    ],
                    "returned": 3,
                  },
                  "stories": Object {
                    "available": 21,
                    "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/stories",
                    "items": Array [
                      Object {
                        "name": "Cover #19947",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/19947",
                        "type": "cover",
                      },
                      Object {
                        "name": "The 3-D Man!",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/19948",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "Cover #19949",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/19949",
                        "type": "cover",
                      },
                      Object {
                        "name": "The Devil's Music!",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/19950",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "Cover #19951",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/19951",
                        "type": "cover",
                      },
                      Object {
                        "name": "Code-Name:  The Cold Warrior!",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/19952",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "AVENGERS: THE INITIATIVE (2007) #14",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/47184",
                        "type": "cover",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #14 - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/47185",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "AVENGERS: THE INITIATIVE (2007) #15",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/47498",
                        "type": "cover",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #15 - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/47499",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "AVENGERS: THE INITIATIVE (2007) #16",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/47792",
                        "type": "cover",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #16 - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/47793",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "AVENGERS: THE INITIATIVE (2007) #17",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/48361",
                        "type": "cover",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #17 - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/48362",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "AVENGERS: THE INITIATIVE (2007) #18",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/49103",
                        "type": "cover",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #18 - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/49104",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #18, Zombie Variant - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/49106",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "AVENGERS: THE INITIATIVE (2007) #19",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/49888",
                        "type": "cover",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #19 - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/49889",
                        "type": "interiorStory",
                      },
                      Object {
                        "name": "Avengers: The Initiative (2007) #14, Spotlight Variant - Int",
                        "resourceURI": "http://gateway.marvel.com/v1/public/stories/54371",
                        "type": "interiorStory",
                      },
                    ],
                    "returned": 20,
                  },
                  "thumbnail": Object {
                    "extension": "jpg",
                    "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
                  },
                  "urls": Array [
                    Object {
                      "type": "detail",
                      "url": "http://marvel.com/characters/74/3-d_man?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                    },
                    Object {
                      "type": "wiki",
                      "url": "http://marvel.com/universe/3-D_Man_(Chandler)?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                    },
                    Object {
                      "type": "comiclink",
                      "url": "http://marvel.com/comics/characters/1011334/3-d_man?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                    },
                  ],
                }
              }
              text="3-D Man"
              type="detail"
            />
            <FavButton
              characterId={1011334}
              isFav={false}
              onFavClicked={[MockFunction]}
            />
          </h5>
          <h6>
            Description
          </h6>
          <p
            className="card-text text-danger"
          />
          <div
            className="row"
          >
            <div
              className="col-sm-6"
            >
              <h6>
                Comics
                <LinkType
                  character={
                    Object {
                      "comics": Object {
                        "available": 12,
                        "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/comics",
                        "items": Array [
                          Object {
                            "name": "Avengers: The Initiative (2007) #14",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/21366",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #14 (SPOTLIGHT VARIANT)",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/24571",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #15",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/21546",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #16",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/21741",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #17",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/21975",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #18",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/22299",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #18 (ZOMBIE VARIANT)",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/22300",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #19",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/22506",
                          },
                          Object {
                            "name": "Deadpool (1997) #44",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/8500",
                          },
                          Object {
                            "name": "Marvel Premiere (1972) #35",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/10223",
                          },
                          Object {
                            "name": "Marvel Premiere (1972) #36",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/10224",
                          },
                          Object {
                            "name": "Marvel Premiere (1972) #37",
                            "resourceURI": "http://gateway.marvel.com/v1/public/comics/10225",
                          },
                        ],
                        "returned": 12,
                      },
                      "description": "",
                      "events": Object {
                        "available": 1,
                        "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/events",
                        "items": Array [
                          Object {
                            "name": "Secret Invasion",
                            "resourceURI": "http://gateway.marvel.com/v1/public/events/269",
                          },
                        ],
                        "returned": 1,
                      },
                      "id": 1011334,
                      "modified": "2014-04-29T14:18:17-0400",
                      "name": "3-D Man",
                      "resourceURI": "http://gateway.marvel.com/v1/public/characters/1011334",
                      "series": Object {
                        "available": 3,
                        "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/series",
                        "items": Array [
                          Object {
                            "name": "Avengers: The Initiative (2007 - 2010)",
                            "resourceURI": "http://gateway.marvel.com/v1/public/series/1945",
                          },
                          Object {
                            "name": "Deadpool (1997 - 2002)",
                            "resourceURI": "http://gateway.marvel.com/v1/public/series/2005",
                          },
                          Object {
                            "name": "Marvel Premiere (1972 - 1981)",
                            "resourceURI": "http://gateway.marvel.com/v1/public/series/2045",
                          },
                        ],
                        "returned": 3,
                      },
                      "stories": Object {
                        "available": 21,
                        "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/stories",
                        "items": Array [
                          Object {
                            "name": "Cover #19947",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/19947",
                            "type": "cover",
                          },
                          Object {
                            "name": "The 3-D Man!",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/19948",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "Cover #19949",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/19949",
                            "type": "cover",
                          },
                          Object {
                            "name": "The Devil's Music!",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/19950",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "Cover #19951",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/19951",
                            "type": "cover",
                          },
                          Object {
                            "name": "Code-Name:  The Cold Warrior!",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/19952",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "AVENGERS: THE INITIATIVE (2007) #14",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/47184",
                            "type": "cover",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #14 - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/47185",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "AVENGERS: THE INITIATIVE (2007) #15",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/47498",
                            "type": "cover",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #15 - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/47499",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "AVENGERS: THE INITIATIVE (2007) #16",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/47792",
                            "type": "cover",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #16 - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/47793",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "AVENGERS: THE INITIATIVE (2007) #17",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/48361",
                            "type": "cover",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #17 - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/48362",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "AVENGERS: THE INITIATIVE (2007) #18",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/49103",
                            "type": "cover",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #18 - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/49104",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #18, Zombie Variant - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/49106",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "AVENGERS: THE INITIATIVE (2007) #19",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/49888",
                            "type": "cover",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #19 - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/49889",
                            "type": "interiorStory",
                          },
                          Object {
                            "name": "Avengers: The Initiative (2007) #14, Spotlight Variant - Int",
                            "resourceURI": "http://gateway.marvel.com/v1/public/stories/54371",
                            "type": "interiorStory",
                          },
                        ],
                        "returned": 20,
                      },
                      "thumbnail": Object {
                        "extension": "jpg",
                        "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
                      },
                      "urls": Array [
                        Object {
                          "type": "detail",
                          "url": "http://marvel.com/characters/74/3-d_man?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                        },
                        Object {
                          "type": "wiki",
                          "url": "http://marvel.com/universe/3-D_Man_(Chandler)?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                        },
                        Object {
                          "type": "comiclink",
                          "url": "http://marvel.com/comics/characters/1011334/3-d_man?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                        },
                      ],
                    }
                  }
                  text=" (12)"
                  type="comiclink"
                />
              </h6>
              <li
                key="Avengers: The Initiative (2007) #14"
              >
                Avengers: The Initiative (2007) #14
              </li>
              <li
                key="Avengers: The Initiative (2007) #14 (SPOTLIGHT VARIANT)"
              >
                Avengers: The Initiative (2007) #14 (SPOTLIGHT VARIANT)
              </li>
              <li
                key="Avengers: The Initiative (2007) #15"
              >
                Avengers: The Initiative (2007) #15
              </li>
              <li
                key="Avengers: The Initiative (2007) #16"
              >
                Avengers: The Initiative (2007) #16
              </li>
              <LinkType
                character={
                  Object {
                    "comics": Object {
                      "available": 12,
                      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/comics",
                      "items": Array [
                        Object {
                          "name": "Avengers: The Initiative (2007) #14",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/21366",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #14 (SPOTLIGHT VARIANT)",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/24571",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #15",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/21546",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #16",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/21741",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #17",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/21975",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #18",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/22299",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #18 (ZOMBIE VARIANT)",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/22300",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #19",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/22506",
                        },
                        Object {
                          "name": "Deadpool (1997) #44",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/8500",
                        },
                        Object {
                          "name": "Marvel Premiere (1972) #35",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/10223",
                        },
                        Object {
                          "name": "Marvel Premiere (1972) #36",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/10224",
                        },
                        Object {
                          "name": "Marvel Premiere (1972) #37",
                          "resourceURI": "http://gateway.marvel.com/v1/public/comics/10225",
                        },
                      ],
                      "returned": 12,
                    },
                    "description": "",
                    "events": Object {
                      "available": 1,
                      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/events",
                      "items": Array [
                        Object {
                          "name": "Secret Invasion",
                          "resourceURI": "http://gateway.marvel.com/v1/public/events/269",
                        },
                      ],
                      "returned": 1,
                    },
                    "id": 1011334,
                    "modified": "2014-04-29T14:18:17-0400",
                    "name": "3-D Man",
                    "resourceURI": "http://gateway.marvel.com/v1/public/characters/1011334",
                    "series": Object {
                      "available": 3,
                      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/series",
                      "items": Array [
                        Object {
                          "name": "Avengers: The Initiative (2007 - 2010)",
                          "resourceURI": "http://gateway.marvel.com/v1/public/series/1945",
                        },
                        Object {
                          "name": "Deadpool (1997 - 2002)",
                          "resourceURI": "http://gateway.marvel.com/v1/public/series/2005",
                        },
                        Object {
                          "name": "Marvel Premiere (1972 - 1981)",
                          "resourceURI": "http://gateway.marvel.com/v1/public/series/2045",
                        },
                      ],
                      "returned": 3,
                    },
                    "stories": Object {
                      "available": 21,
                      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/stories",
                      "items": Array [
                        Object {
                          "name": "Cover #19947",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/19947",
                          "type": "cover",
                        },
                        Object {
                          "name": "The 3-D Man!",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/19948",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "Cover #19949",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/19949",
                          "type": "cover",
                        },
                        Object {
                          "name": "The Devil's Music!",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/19950",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "Cover #19951",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/19951",
                          "type": "cover",
                        },
                        Object {
                          "name": "Code-Name:  The Cold Warrior!",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/19952",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "AVENGERS: THE INITIATIVE (2007) #14",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/47184",
                          "type": "cover",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #14 - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/47185",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "AVENGERS: THE INITIATIVE (2007) #15",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/47498",
                          "type": "cover",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #15 - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/47499",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "AVENGERS: THE INITIATIVE (2007) #16",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/47792",
                          "type": "cover",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #16 - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/47793",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "AVENGERS: THE INITIATIVE (2007) #17",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/48361",
                          "type": "cover",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #17 - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/48362",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "AVENGERS: THE INITIATIVE (2007) #18",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/49103",
                          "type": "cover",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #18 - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/49104",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #18, Zombie Variant - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/49106",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "AVENGERS: THE INITIATIVE (2007) #19",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/49888",
                          "type": "cover",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #19 - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/49889",
                          "type": "interiorStory",
                        },
                        Object {
                          "name": "Avengers: The Initiative (2007) #14, Spotlight Variant - Int",
                          "resourceURI": "http://gateway.marvel.com/v1/public/stories/54371",
                          "type": "interiorStory",
                        },
                      ],
                      "returned": 20,
                    },
                    "thumbnail": Object {
                      "extension": "jpg",
                      "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
                    },
                    "urls": Array [
                      Object {
                        "type": "detail",
                        "url": "http://marvel.com/characters/74/3-d_man?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                      },
                      Object {
                        "type": "wiki",
                        "url": "http://marvel.com/universe/3-D_Man_(Chandler)?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                      },
                      Object {
                        "type": "comiclink",
                        "url": "http://marvel.com/comics/characters/1011334/3-d_man?utm_campaign=apiRef&utm_source=fc0bbfff7046baab7764e1ef9a394a25",
                      },
                    ],
                  }
                }
                text="Details"
                type="comiclink"
              />
            </div>
            <div
              className="col-sm-6"
            >
              <h6>
                Series
              </h6>
              <li
                key="Avengers: The Initiative (2007 - 2010)"
              >
                Avengers: The Initiative (2007 - 2010)
              </li>
              <li
                key="Deadpool (1997 - 2002)"
              >
                Deadpool (1997 - 2002)
              </li>
              <li
                key="Marvel Premiere (1972 - 1981)"
              >
                Marvel Premiere (1972 - 1981)
              </li>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
